src/limb.rs

```diff
@@ -0,0 +1,220 @@
+use {rand, polyfill, c, core, error};
+
+// XXX: Not correct for x32 ABIs.
+#[cfg(target_pointer_width = "64")] pub type Limb = u64;
+#[cfg(target_pointer_width = "32")] pub type Limb = u32;
+#[cfg(target_pointer_width = "64")] pub const LIMB_BITS: usize = 64;
+#[cfg(target_pointer_width = "32")] pub const LIMB_BITS: usize = 32;
+
+pub const LIMB_BYTES: usize = (LIMB_BITS + 7) / 8;
+
+/// References a positive integer range `[1..max_exclusive)` where
+/// `max_exclusive` is a slice of limbs ordered least-significant-limb to
+/// most-significant-limb (and the limbs use the native endianness).
+pub struct Range<'a> {
+    pub max_exclusive: &'a [Limb],
+}
+
+impl <'a>Range<'a> {
+
+    pub fn from_max_exclusive(max_exclusive: &[Limb]) -> Range {
+
+        debug_assert!(!limbs_are_zero_constant_time(max_exclusive));
+        debug_assert!(max_exclusive[max_exclusive.len() - 1] > 0);
+
+        Range {
+            max_exclusive: max_exclusive
+        }
+    }
+
+    fn leading_zero_bits(&self) -> usize {
```

Comment:
Huh, I googled for something like these and didn't have any luck. I guess I should have just scanned the methods on u32/64 instead.